interactive_templates/create.py

```diff
@@ -61,7 +62,9 @@ def commit_and_push(working_dir, analysis, force=False):
 
 
 def get_repo_with_token(repo, token):
-    if repo.startswith("https://github.com"):
+    url = urlparse(repo)
+
+    if url.hostname == "github.com" and url.scheme == "https":
```

Comment:
thought: I think this is probably OK for what it's doing.

Unfortunately, there's no easy way to replace the username and password in a parsed URL with `urllib.parse`.

Reply:
not quite sure what you mean here

```python
>>> u = urlparse("https://foo:bar@flibble.com/one/two?three")
>>> u.username
'foo'
>>> u.password
'bar'
>>> u.hostname
'flibble.com'
```

Reply:
Oh, I see, instead of the following string replace.

Reply:
something like
```python
>>> u = urlparse("https://github.com/opensafely/foo")
>>> token = "abcd"
>>> u = u._replace(netloc=f"interactive:{token}@{u.netloc}")
>>> u.geturl()
'https://interactive:abcd@github.com/opensafely/foo'
```

but I'm not 100% what that would gain us

Reply:
unfortunately, `username` and `password` aren't modifiable via the `_replace()` method

Reply:
> unfortunately, `username` and `password` aren't modifiable via the `_replace()` method

Yes, that's what I was referring to.

Reply:
Do you have any thoughts on my most recent pushes which added some tests and a (potentially) safer reconstruction of the URL from its parsed components?

Reply:
I found this function hard to reason about: it took me awhile to realise that we weren't using the parse result object (returned by `urlparse`) for updating `repo`. How about...

```py
def get_repo_with_token(repo, token):
    scheme, netloc, path, params, query, fragment = urlparse(repo)
    if scheme == "https" and netloc == "github.com":
        new_netloc = f"interactive:{token}@"
        return urlunparse((scheme, new_netloc, path, params, query, fragment))

    return repo
```

Here, I've intentionally not modified `repo` and I've separated the section that makes use of the parts from the noop section. I've also reordered `scheme` and `netloc` in `if`, such that they match the order in which they are unpacked.